### PR TITLE
[Type Refactor] Merge TType and Tensor instances as a single entity

### PR DESCRIPTION
**Pull request by karllessard**
This PR introduces some breaking changes related to `Tensor` and their relationship with the `NdArray` data structure, in an effort of simplification and standardization.

Instances of `TType` (e.g. `TFloat32`, `TInt32`, etc.) now implements directly `Tensor`, which has been converted to a non-parameterized interface. This allows direct access to the tensor n-dimensional data space from a tensor instance, without the need to dereference to a second object by calling `Tensor.data()`.

```java
// Before
Tensor<TFloat32> t = TFloat32.vectorOf(0, 1, 2, 3);
TFloat32 data = t.data();
data.getFloat(1); // 1.0f

// Now
TFloat32 t = TFloat32.vectorOf(0, 1, 2, 3);
t.getFloat(1); // 1.0f
```

An instance of `TType` can now be passed directly to any method accepting a `Tensor` in input, e.g. `Session.Runner.feed`. In addition, it is now possible to cast directly a tensor of an unknown type using standard Java explicit or implicit type casting, instead of the custom method `expect`.

```java
// Before
Tensor<TFloat32> t = session.run().get(0).expect(TFloat32.DTYPE);

// Now
TFloat32 = (TFloat32)s.run().get(0);
```

In a nutshell, important and/or breaking changes are:

* Convert `Tensor` from a parameterized class to an non-parameterized interface
* Implement directly `Tensor` from `TType` implementations
* Add a `RawTensor` implementation of `Tensor` that takes care, among other things, of wrapping the native handle of the tensor in a protected way
* Drop now obsolete `Tensor.data()`, `Tensor.expect()` and `Operand.data()` methods
* `tf.constant(TType)` has to be renamed to `tf.capture(TType)` to avoid conflict with other `Constant` endpoints (left untouched).
    * In a next PR, the use of `tf.capture` will also become obsolete by implementing directly `Operand` from tensors.

CC\ @deansher , @Craigacp , @wmeddie

### CODE REVIEW

**Review by deansher — CHANGES_REQUESTED** [review]

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/DataType.java [hunk 1]
```diff
@@ -35,13 +34,13 @@
...
      *
-     * @param nativeTensor pointer to the native tensor
-     * @param shape the shape of the tensor
-     * @return data structure of elements of this type
+     * @param tensor the tensor to map in its raw nature
+     * @param nativeHandle native handle of the tensor
+     * @return a typed tensor of type {@code T}
      */
-    T apply(TF_Tensor nativeTensor, Shape shape);
+    T apply(RawTensor tensor, TF_Tensor nativeHandle);
```
[the remark this answers is not in the document]
Comment (karllessard) [reply tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/DataType.java:43]: The reason I'm passing `nativeHandle` as a separate parameter here is because I want to keep `RawTensor.nativeHandle()` package-private, reserving its use to what is located under `org.tensorflow` only. But as implementors of `TensorMapper` are outside this package, `apply` serves also as a visitor to `RawTensor` internal representation.

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/EagerOperation.java [hunk 2]
```diff
@@ -156,13 +156,13 @@ private static void requireTensorHandle(TFE_TensorHandle handle) {
...
 
-  private static Tensor<?> resolveTensorHandle(TFE_TensorHandle handle, EagerSession session) {
+  private static Tensor resolveTensorHandle(TFE_TensorHandle handle, EagerSession session) {
     requireTensorHandle(handle);
     try (PointerScope scope = new PointerScope()) {
       TF_Status status = TF_Status.newStatus();
       TF_Tensor tensor = TFE_TensorHandleResolve(handle, status).withDeallocator();
       status.throwExceptionIfNotOK();
-      return Tensor.fromHandle(tensor, session);
+      return RawTensor.fromHandle(tensor, session).asTypedTensor();
```
[the remark this answers is not in the document]
Comment (karllessard) [reply tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/EagerOperation.java:165]: `RawTensor.asTypedTensor()` is also a package-private method for typing a tensor from its raw nature. The reason why I'm doing it here is that tensors resolved in an eager operation are cached in that object and are ultimately typed before they can be used by the user. Meaning that if we only preserve a raw copy of the tensor in cache, then they will need to be converted as a typed tensor everytime the user get accesses them. 

Therefore, typing them right away at this level is simply to improve performances when the tensors are accessed multiple times via its container (e.g. `Operand.asTensor()).

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/types/TBool.java [hunk 3]
```diff
@@ -111,19 +111,33 @@
...
   }
 }
 
 /** Hidden implementation of a {@code TBool} */
 class TBoolImpl extends BooleanDenseNdArray implements TBool {
 
-  static TBool mapTensor(TF_Tensor nativeTensor, Shape shape) {
-    return new TBoolImpl(TensorBuffers.toBooleans(nativeTensor), shape);
+  @Override
+  public DataType<?> dataType() {
```
[the remark this answers is not in the document]
Comment (karllessard) [reply tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/types/TBool.java:123]: Based on how the tensors are converted from their raw nature to their typed one, I don't think it is possible that both values end up being out-of-sync. But anyhow, the ultimate plan is to return `TBool.class` and not `TBool.DTYPE` in `TBoolImpl`, while in `RawTensor` the datatype will be returned as a raw value instead, as defined in [this enumeration](https://github.com/tensorflow/java/blob/master/tensorflow-core/tensorflow-core-api/src/gen/java/org/tensorflow/proto/framework/DataType.java). 

So maybe we can mitigate this detail until we complete the changes and split `dataType()` in 2 different methods? I was thinking that the method returning a type class like `TBool.class` could be declared in `TType`  (maybe called `type()` or `tensorType()`) while `dataType()` will remain in `Tensor` but returning an enum value instead.

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/types/family/TType.java [hunk 4]
```diff
@@ -17,21 +17,40 @@
...
+ * Constant<TFloat32> c1 = tf.array(2.0f, 3.0f, 2.0f);
+ * Constant<TFloat32> c2 = tf.array(1.0f, 2.0f, 3.0f);
+ * Constant<TInt32> c3 = tf.array(2, 3, 2);
+ *
+ * tf.math.add(c1, c2);  // OK
+ * tf.math.add(c1, c3);  // Compilation failure
  * }</pre>
  */
-public interface TType {}
+public interface TType extends Tensor {
```
[the remark this answers is not in the document]
Comment (karllessard) [reply tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/types/family/TType.java:45]: I deliberately tried to leave `TType` unparameterized, partially to reduce verbosity since this interface is carried in many endpoints and containers. If we add a parameter to `TType`, signatures with `T extends TType` now become `T extends TType<U>, U`, where you'll need to carry `U` everywhere as well, unless you turn it into a wildcard but then you lose the plus value of having a parameter.

Both `TType` and `RawTensor` instances are a `Tensor`, but only the former is typed so that a user can access its data properly but requiring him to cast it to a concrete type to do it. This is intentional as it leverages the use of primitive access endpoints to retrieve the data instead of the boxed ones. 

For example, if a user has an instance of `TType<Float>` that extends from `NdArray<Float>`, chances are that he would retrieve its data probably directly via the boxed-type methods like `t.getObject()`, which would return costly `Float` objects. 

But by enforcing a cast from `TType` to `TFloat32` instead, he would retrieve the same data but now being able to use `t.getFloat()` from `FloatNdArray`, which returns `float` primitive values, improving performances overall.

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/DataType.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (deansher) [reply tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/DataType.java:43]: Convincing! Perhaps incorporate this explanation as a comment?

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/EagerOperation.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (deansher) [reply tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/EagerOperation.java:165]: Convincing! Perhaps incorporate this explanation as a comment?

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/types/family/TType.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (deansher) [reply tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/types/family/TType.java:45]: Convincing! Perhaps incorporate the essence of this explanation as a comment?

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/types/TBool.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (deansher) [reply tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/types/TBool.java:123]: I am comfortable with this direction, and with leaving it this way for now.

File: tensorflow-framework/src/main/java/org/tensorflow/framework/utils/ShapeUtils.java [hunk 5]
```diff
@@ -63,64 +63,36 @@
...
-    return data.shape();
+  public static long[] getLongArray(Tensor dims) {
+    List<Long> result = new ArrayList<>();
+    if (dims instanceof TInt32) {
+      ((TInt32)dims).scalars().forEach(s -> result.add((long) s.getInt()));
+    } else if (dims instanceof TInt64) {
+      ((TInt64)dims).scalars().forEach(s -> result.add(s.getLong()));
+    } else if (dims instanceof TUint8) {
+      ((TUint8)dims).scalars().forEach(s -> result.add(s.getObject().longValue()));
+      } else { // shouldn't happen
```
Comment (Craigacp) [inline tensorflow-framework/src/main/java/org/tensorflow/framework/utils/ShapeUtils.java:92]: Formatting seems off in this else block. Also maybe we should have a `TIntegral` type to go with `TFloating` and then the else block really can't happen. And when we get sealed types in a future Java version we can have the compiler check it for us.

**Review by Craigacp — CHANGES_REQUESTED** [review]
While we're in here, is it worth moving the contents of the `types.family` package into the `types` package? It seems strange to me to have the interfaces at a lower level than the concrete implementations.

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/DataType.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (karllessard) [reply tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/DataType.java:43]: > We could add a default method that accepts a RawTensor and unpacks it into the RawTensor and the TF_Tensor and calls apply

Current signature for `apply` also ensure that only classes of the `org.tensorflow` package can invoke it, since you shouldn't be able to have a reference to a tensor native handle. Just accepting `RawTensor` will open it up to the outside world. Now, let's not be zealous about securing the tensor native handle neither, I think this is a good trade off, I'll make the changes.

File: tensorflow-core/tensorflow-core-api/src/gen/annotations/org/tensorflow/op/DebuggingOps.java [hunk 6]
```diff
@@ -1,50 +0,0 @@
...
-   *  <p>
-   *  When run, reports an `InvalidArgument` error if `tensor` has any values
-   *  that are not a number (NaN) or infinity (Inf). Otherwise, passes `tensor` as-is.
-   *
-   * @param <T> data type for {@code output()} output
-   * @param tensor
-   * @param message Prefix of the error message.
-   * @return a new instance of CheckNumerics
-   */
-  public <T extends TNumber> CheckNumerics<T> checkNumerics(Operand<T> tensor, String message) {
```
[the remark this answers is not in the document]
Comment (karllessard) [reply tensorflow-core/tensorflow-core-api/src/gen/annotations/org/tensorflow/op/DebuggingOps.java:47]: I think it has been removed from the TF runtime library and we have just never updated it (ideally, we should delete our generated file folder before regenerating them, I think right now deleted ops remain present). I will need to double-check.

File: tensorflow-core/tensorflow-core-api/src/gen/annotations/org/tensorflow/op/Ops.java [hunk 7]
```diff
@@ -1071,6 +1070,17 @@ public Batch batch(Iterable<Operand<?>> inTensors, Long numBatchThreads, Long ma
...
   }
 
+  /**
+   * Capture a {@code tensor} by making a constant copy of it.
+   *
+   * @param scope is a scope used to add the underlying operation.
+   * @param tensor a Tensor holding the constant value
+   * @return a constant of the same data type as `tensor`
+   */
+  public <T extends TType> Constant<T> capture(T tensor) {
```
[the remark this answers is not in the document]
Comment (karllessard) [reply tensorflow-core/tensorflow-core-api/src/gen/annotations/org/tensorflow/op/Ops.java:1080]: I don't like capture neither. As discussed [here](https://github.com/tensorflow/java/pull/166#issuecomment-741788540), eventually the tensors could be automatically converted to constants so we will not need to call this endpoint anymore. Still, I now think `copy` do sound more appropriate as well, I'll make the change.

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/RawTensor.java [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+import org.bytedeco.javacpp.PointerScope;
+import org.tensorflow.internal.buffer.TensorBuffers;
+import org.tensorflow.internal.c_api.TF_Tensor;
+import org.tensorflow.ndarray.Shape;
+import org.tensorflow.ndarray.Shaped;
+import org.tensorflow.ndarray.buffer.ByteDataBuffer;
+import org.tensorflow.types.family.TType;
+
+/**
+ * A tensor which memory has not been mapped.
```
[the remark this answers is not in the document]
Comment (karllessard) [reply tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/RawTensor.java:32]: A tensor never resides on the Java heap, it is always in native memory and until we map it to a given `TType`, we can only access its data as a flat buffer of bytes.

So `mapped` here means `mapped to a multidimensional data space directly accessible from the JVM`, want me to add something like this?

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Tensor.java [hunk 9]
```diff
@@ -92,19 +85,13 @@
...
-    TF_Tensor nativeHandle = allocate(t.dtype.nativeCode(), shape.asArray(), size);
-    try (PointerScope scope = new PointerScope()) {
-        scope.attach(nativeHandle);
-        t.tensorHandle = nativeHandle;
-        t.tensorScope = scope.extend();
-        return t;
+  static <T extends TType> T of(DataType<T> dtype, Shape shape, long size) {
+    RawTensor tensor = RawTensor.allocate(dtype, shape, size);
+    try {
+      return dtype.map(tensor);
```
[the remark this answers is not in the document]
Comment (karllessard) [reply tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Tensor.java:91]: Ah probably I could, I've added `asTypedTensor()` later in my work

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Tensor.java [hunk 10]
```diff
@@ -182,203 +167,39 @@
...
-    }
-    return numBytes;
-  }
-
-  /** Returns the shape of this tensor. */
-  @Override
-  public Shape shape() {
-    return shape;
-  }
+  DataType<?> dataType();
```
[the remark this answers is not in the document]
Comment (karllessard) [reply tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Tensor.java:177]: The next PR will replace `DataType<?>` by [`DataType`](https://github.com/karllessard/tensorflow-java/blob/tensor-as-ndarrays/tensorflow-core/tensorflow-core-api/src/gen/java/org/tensorflow/proto/framework/DataType.java), among other things. I was trying to limit the number of changes in this one.

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/types/TBfloat16.java [hunk 11]
```diff
@@ -119,20 +120,33 @@
...
 }
 
 /** Hidden implementation of a {@code TBfloat16} */
 class TBfloat16Impl extends FloatDenseNdArray implements TBfloat16 {
 
-  static TBfloat16 mapTensor(TF_Tensor nativeTensor, Shape shape) {
-    return new TBfloat16Impl(
-        DataLayouts.BFLOAT16.applyTo(TensorBuffers.toShorts(nativeTensor)), shape);
+  @Override
+  public DataType<?> dataType() {
```
[the remark this answers is not in the document]
Comment (karllessard) [reply tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/types/TBfloat16.java:132]: See https://github.com/tensorflow/java/pull/160#discussion_r539344901

File: tensorflow-core/tensorflow-core-api/src/test/java/org/tensorflow/op/core/GradientsTest.java [hunk 12]
```diff
@@ -47,13 +47,13 @@ public void createGradients() {
       assertNotNull(grads.dy());
       assertEquals(2, grads.dy().size());
 
-      try (Tensor<TFloat32> c = TFloat32.scalarOf(3.0f);
-          AutoCloseableList<Tensor<?>> outputs =
+      try (TFloat32 c = TFloat32.scalarOf(3.0f);
+          AutoCloseableList<?> outputs =
```
[the remark this answers is not in the document]
Comment (karllessard) [reply tensorflow-core/tensorflow-core-api/src/test/java/org/tensorflow/op/core/GradientsTest.java:51]: Yes, looks like a massive auto-replace mistake, thanks!

File: tensorflow-framework/src/main/java/org/tensorflow/framework/utils/ShapeUtils.java (hunk 5 above)
Comment (karllessard) [reply tensorflow-framework/src/main/java/org/tensorflow/framework/utils/ShapeUtils.java:92]: `TIntegral` is also planned, I was keeping this for another PR as well, as there is more functions in the framework that will need to be updated.

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/DataType.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (Craigacp) [reply tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/DataType.java:43]: Sure, that's a good point. If it's something that's designed to be locked off when we eventually have modules then maybe we should just document it that way and move on.

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/RawTensor.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (Craigacp) [reply tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/RawTensor.java:32]: Sure.

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Tensor.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (Craigacp) [reply tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Tensor.java:177]: Sure, I lost sight of the overall goal for a moment.

**Comment by karllessard** [comment]
> While we're in here, is it worth moving the contents of the `types.family` package into the `types` package? It seems strange to me to have the interfaces at a lower level than the concrete implementations.

I'm not opposed to this but let's keep it for another PR, more changes in this package are planned for the refactoring in progress.

File: tensorflow-core/tensorflow-core-api/src/gen/annotations/org/tensorflow/op/Ops.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (karllessard) [reply tensorflow-core/tensorflow-core-api/src/gen/annotations/org/tensorflow/op/Ops.java]: Ok as `copy` already exist as another primitive op, I've decided to rename `capture` by `constantOf`, as it is close enough to the original `constant` name and will be easier to find by our users on auto-completion.

**Review by Craigacp — APPROVED** [review]
LGTM
